### PR TITLE
feat: make CLI deployment default, gate API behind experimental-api flag

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Restore the Vercel CLI as the default deployment client and gate the `@vercel/client` API path behind a new `experimental-api` opt-in input. The API client depends on an internal Vercel package without semver guarantees, so users should opt in consciously.

## Track

- Track: [`cli-default-experimental-api-20260430`](./.please/docs/tracks/completed/cli-default-experimental-api-20260430/)
- Spec: [spec.md](./.please/docs/tracks/completed/cli-default-experimental-api-20260430/spec.md)
- Plan: [plan.md](./.please/docs/tracks/completed/cli-default-experimental-api-20260430/plan.md)
- Refs: #362

## Routing matrix

| `experimental-api` | `vercel-args` | Result |
|---|---|---|
| false (default) | empty | `VercelCliClient` |
| false (default) | `--prod` | `VercelCliClient` |
| true | empty | `VercelApiClient` (with experimental warning) |
| true | `--prod` | Config error (mutually exclusive, thrown at config-parse time) |

## Type-level invariant

After the review-driven refactor, the mutual-exclusion invariant is encoded in the type system:

```ts
type DeploymentMode =
  | { kind: 'cli', vercelArgs: string }
  | { kind: 'experimental-api' }

interface ActionConfig {
  deployment: DeploymentMode
  // ...
}
```

The `(experimental-api: true, vercel-args: '...')` combination is unrepresentable. The runtime check in `parseDeploymentMode()` exists only to validate raw action.yml inputs at the boundary.

## Implementation Tasks

- [x] T001 Add experimental-api input
- [x] T002 ActionConfig type (later refactored into DeploymentMode union)
- [x] T003 Parse + mutual exclusion
- [x] T004 Config tests
- [x] T005 Routing change + warning
- [x] T006 Routing matrix tests
- [x] T007 Test helper update
- [x] T008 action.yml cleanup
- [x] T009 README Deployment Mode section
- [x] T010 README migration note
- [x] T011 Integration tests
- [x] T012 Quality gate

## Verification Checklist

- [ ] `pnpm lint` — no warnings (verified locally: 0 errors, 20 pre-existing warnings)
- [ ] `pnpm typecheck` — passes (verified locally)
- [ ] `pnpm test` — all unit tests pass, including the four-case routing matrix (243/243 passing)
- [ ] `pnpm test:integration` — emulator tests pass for both default (CLI) and `experimental-api: true` (API) paths (19/19 passing)
- [ ] `pnpm build` — `dist/index.js` regenerated successfully (committed)
- [ ] Manual: invoke the action with default inputs and confirm `core.info('Using CLI-based deployment')` is logged (no warning)
- [ ] Manual: invoke the action with `experimental-api: true` and confirm `core.warning(...)` is logged exactly once
- [ ] Manual: invoke the action with both `experimental-api: true` and `vercel-args: --prod` and confirm a clear config error is thrown before any deployment side effects

## Review

- Spec compliance: 17/17 IMPLEMENTED
- Code review: 0 issues at confidence ≥80 (iteration 2 clean)
- Type review: original suggestion applied via discriminated union refactor
- Test review: AC-1 four-case matrix and AC-6 single-warning assertion verified

🤖 Created via /please:implement, refined via /please:review

### CODE REVIEW

File: .please/docs/tracks/completed/cli-default-experimental-api-20260430/spec.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+## Requirements
+
+### Functional Requirements
+
+- [ ] FR-1: Add a new `experimental-api` boolean input to `action.yml` (default: `false`).
+- [ ] FR-2: Default routing in `createVercelClient()` returns `VercelCliClient` when `experimental-api` is `false` (or unset).
+- [ ] FR-3: When `experimental-api` is `true`, `createVercelClient()` returns `VercelApiClient` and emits a `core.warning` line stating that API mode is experimental and may break across `@vercel/client` updates.
+- [ ] FR-4: When `experimental-api` is `true` AND `vercel-args` is non-empty, the action MUST fail fast with a clear configuration error explaining that the two inputs are mutually exclusive, and instruct the user to choose one.
+- [ ] FR-5: When `vercel-args` is non-empty and `experimental-api` is `false`, route to `VercelCliClient` (legacy passthrough preserved).
+- [ ] FR-6: Remove the `deprecationMessage` block from `vercel-args` and `scope` inputs in `action.yml` — CLI is now the standard mode and these inputs are no longer deprecated.
```
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/tracks/completed/cli-default-experimental-api-20260430/spec.md:24]: ⏭️ Not applied — the spec lives in the completed/ track and reflects the originally approved requirement; the actual implementation already addressed the inconsistency you flagged (the deprecationMessage was removed from scope per spec literal, while vercel-args got a real description added). The mismatch between FR-6's wording and the real action.yml state is documented in plan.md's 'Surprises & Discoveries' section. Updating the spec post-finalize would be revisionism on an already-merged track artifact.

Thanks for catching the wording inconsistency, @gemini-code-assist!

File: .please/docs/tracks/active/cli-default-experimental-api-20260430/plan.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+### Phase 1 — Config foundation
+
+- [ ] T001 Add `experimental-api` boolean input to `action.yml` with default `'false'` and a description explaining it gates opt-in API mode and is mutually exclusive with `vercel-args` (file: action.yml)
+- [ ] T002 Add `experimentalApi: boolean` field to `ActionConfig` interface (file: src/types.ts) (depends on T001)
+- [ ] T003 Parse the new input in `getActionConfig()` and add the mutual-exclusion error: throw a clear `Error` when both `experimental-api === 'true'` and `vercel-args` is non-empty (file: src/config.ts) (depends on T002)
+- [ ] T004 Add unit tests in `src/__tests__/config.test.ts` for: (a) `experimentalApi=false` default, (b) `experimentalApi=true` parses correctly, (c) mutual-exclusion error message and stack when both are set (file: src/__tests__/config.test.ts) (depends on T003)
+
+### Phase 2 — Routing change
+
+- [ ] T005 Update `createVercelClient()` in `src/vercel.ts`: route to `VercelCliClient` by default; route to `VercelApiClient` only when `config.experimentalApi === true`; emit `core.warning` describing the experimental nature and how to opt out when API path is taken (file: src/vercel.ts) (depends on T003)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/tracks/active/cli-default-experimental-api-20260430/plan.md]: ⏭️ Not applied — the plan lives in the completed/ track and is a historical record of what was planned at the start. The actual implementation already does what you suggest: src/vercel.ts now logs 'Using CLI-based deployment' (no longer mentions vercel-args) on the CLI path, and the routing tests assert that exact string. The plan was followed in spirit; updating it after the fact would be backdating.

Thanks for the careful read, @gemini-code-assist!

File: src/config.ts [hunk 3]
```diff
@@ -92,7 +106,7 @@ export function getActionConfig(): ActionConfig {
     githubDeploymentEnvironment: resolveDeploymentEnvironment(githubDeploymentEnvInput, vercelArgs),
     workingDirectory: parseWorkingDirectory(core.getInput('working-directory')),
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/config.ts:135]: ✅ Applied in [`633901c`](https://github.com/amondnet/vercel-action/pull/363/commits/633901c)

- File: `src/config.ts`
- Change: `resolveDeploymentEnvironment` now takes the parsed `DeploymentMode` and the typed `target`. In `experimental-api` mode the GitHub deployment environment is inferred from `target === 'production'` rather than from `vercel-args` (which is forbidden in that mode).
- Tests: 2 new test cases cover the experimental-api branches; the existing `--prod` regex path still applies in CLI mode.

Thanks for catching this — production API deploys would have been mislabeled as preview, @copilot-pull-request-reviewer!

File: src/vercel.ts [hunk 4]
```diff
@@ -7,12 +7,18 @@ import { VercelCliClient } from './vercel-cli'
...
 
 export function createVercelClient(config: ActionConfig): VercelClient {
-  if (config.vercelArgs) {
-    core.info('Using CLI-based deployment (vercel-args provided)')
-    return new VercelCliClient(config)
+  switch (config.deployment.kind) {
+    case 'experimental-api':
+      core.warning(
+        'Using experimental API-based deployment via @vercel/client. '
+        + 'This is an internal Vercel package without semver guarantees and may break across updates. '
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/vercel.ts:14]: ✅ Applied in [`633901c`](https://github.com/amondnet/vercel-action/pull/363/commits/633901c)

- File: `src/vercel.ts`
- Change: added a `default` branch with an `assertNever`-style exhaustiveness check that throws a clear error if `config.deployment.kind` is ever an unexpected value. Future variants will produce a compile error at the assignment to `never`, and any runtime injection (e.g., from a misshapen JSON) fails loud instead of returning `undefined`.

Thanks for the defensive-coding nudge, @copilot-pull-request-reviewer!

File: src/config.ts [hunk 5]
```diff
@@ -78,11 +78,25 @@ export function resolveDeploymentEnvironment(explicitEnv: string, vercelArgs: st
   return /(?:^|\s)--prod(?:uction)?(?:\s|$)/.test(vercelArgs) ? 'production' : 'preview'
 }
 
+function parseDeploymentMode(vercelArgs: string, experimentalApi: boolean): DeploymentMode {
+  if (experimentalApi && vercelArgs) {
+    throw new Error(
+      'The "experimental-api" and "vercel-args" inputs are mutually exclusive. '
+      + 'Either remove "vercel-args" to use the experimental API mode, '
+      + 'or set "experimental-api: false" (or remove it) to use CLI mode with vercel-args.',
+    )
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/config.ts:95]: ✅ Applied in [`633901c`](https://github.com/amondnet/vercel-action/pull/363/commits/633901c)

- File: `src/config.ts`
- Change: `parseDeploymentMode` now trims `vercel-args` before the mutual-exclusion check, and stores the trimmed value in the `cli` variant. Whitespace-only input (`'   \\t  '`) no longer triggers a spurious mutual-exclusion error and round-trips as the same effectively-empty value that `parseArgs()` would have produced anyway.
- Tests: 2 new tests cover the whitespace-only case (no throw, returns experimental-api) and the trimmed-cli case (`'  --prod  '` → `{ kind: 'cli', vercelArgs: '--prod' }`).

Thanks for catching the edge case, @copilot-pull-request-reviewer!

File: src/config.ts [hunk 6]
```diff
@@ -78,11 +78,25 @@ export function resolveDeploymentEnvironment(explicitEnv: string, vercelArgs: st
...
+  return experimentalApi ? { kind: 'experimental-api' } : { kind: 'cli', vercelArgs }
+}
+
 export function getActionConfig(): ActionConfig {
   const vercelToken = core.getInput('vercel-token', { required: true })
   core.setSecret(vercelToken)
 
   const vercelArgs = core.getInput('vercel-args')
+  const experimentalApi = core.getInput('experimental-api') === 'true'
+  const deployment = parseDeploymentMode(vercelArgs, experimentalApi)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/config.ts:106]: ✅ Applied in [`633901c`](https://github.com/amondnet/vercel-action/pull/363/commits/633901c)

Same fix as the parallel @copilot-pull-request-reviewer thread on the same line. `resolveDeploymentEnvironment` now incorporates the typed `target` when `deployment.kind === 'experimental-api'`, so production API deploys correctly label the GitHub deployment environment as `production` without requiring `github-deployment-environment` to be set manually.

Thanks for catching this in parallel, @cubic-dev-ai!

File: .please/docs/tracks/completed/cli-default-experimental-api-20260430/plan.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+## Progress
+
+- 2026-04-30 — Phase 1 complete: T001 added `experimental-api` input; T002 added `experimentalApi: boolean` to `ActionConfig`; T003 added parsing + mutual-exclusion validation; T004 added 7 config tests (defaults, parsing, error message)
+- 2026-04-30 — Phase 2 complete: T005 inverted routing default in `createVercelClient()` and emits `core.warning` when API path is taken; T006 rewrote `describe('createVercelClient', ...)` as the four-case AC-1 matrix; T007 updated `createConfig` test helper
+- 2026-04-30 — Phase 3 complete: T008 replaced empty `vercel-args` description, reworded `scope` deprecation message
+- 2026-04-30 — Phase 4 complete: T009 replaced "Migration to API-based Deployment" with "Deployment Mode" section in README; T010 added migration note for users on the previous v42 API default
+- 2026-04-30 — Phase 5 complete: T011 added factory-routing integration tests + `experimentalApi` field to integration `createConfig`; T012 quality gate (lint, typecheck, build, full test suite + integration suite) all green — 244 unit + 19 integration tests pass
```
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/tracks/completed/cli-default-experimental-api-20260430/plan.md]: ✅ Applied in [`633901c`](https://github.com/amondnet/vercel-action/pull/363/commits/633901c)

- File: `.please/docs/tracks/completed/cli-default-experimental-api-20260430/plan.md`
- Change: corrected the historical Phase 5 entry from 244 → 243 unit tests (the count dropped by 1 when the discriminated-union refactor merged 2 tests into 1), with a parenthetical note that the count is now 248 after the post-review fixes in this commit.

Thanks for catching the doc drift, @cubic-dev-ai!

File: README.md [hunk 8]
```diff
@@ -50,18 +50,19 @@ This action make a Vercel deployment with github actions.
...
 | vercel-org-id                    | <ul><li>- [x] </li></ol> |         | Vercel team ID (also used as `teamId` for API deployments). See [How can I use GitHub Actions with Vercel](https://vercel.com/kb/guide/how-can-i-use-github-actions-with-vercel) |
-| vercel-args                      | <ul><li>- [ ] </li></ol> |         | ⚠️ **Deprecated.** Use the new API inputs below instead. When provided, falls back to CLI-based deployment. |
+| vercel-args                      | <ul><li>- [ ] </li></ol> |         | Ad-hoc CLI flags forwarded to the Vercel CLI (e.g. `--prod --force`). Mutually exclusive with `experimental-api`. |
 | working-directory                | <ul><li>- [ ] </li></ol> |         | the working directory                                                                             |
-| scope                            | <ul><li>- [ ] </li></ol> |         | ⚠️ **Deprecated.** Team slug for CLI `--scope` flag. For API deployments, use `vercel-org-id` instead.
+| scope                            | <ul><li>- [ ] </li></ol> |         | Team slug for the Vercel CLI `--scope` flag. Prefer `vercel-org-id`, which works in both CLI and experimental API modes.
```
[the remark this answers is not in the document]
Comment (amondnet) [reply README.md]: ✅ Applied in [`633901c`](https://github.com/amondnet/vercel-action/pull/363/commits/633901c)

- File: `README.md`
- Change: added the trailing `|` to the `scope` row of the inputs table so the markdown table parses cleanly.

Thanks, @cubic-dev-ai!
